File: app/src/main/java/com/alpdroid/huGen10/ui/MainActivity.java
package com.alpdroid.huGen10.ui;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.preference.PreferenceManager;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.alpdroid.huGen10.AlpdroidApplication;
import com.alpdroid.huGen10.BuildConfig;
import com.alpdroid.huGen10.CanFrame;
import com.alpdroid.huGen10.OsmAndHelper;
import com.alpdroid.huGen10.R;
import com.alpdroid.huGen10.VehicleServices;
import com.google.android.material.tabs.TabLayout;
import com.google.common.collect.ImmutableList;

import org.osmdroid.config.Configuration;

import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import m.co.rh.id.alogger.ILogger;

public class MainActivity extends AppCompatActivity implements OsmAndHelper.OnOsmandMissingListener {

        private final String TAG = "MainActivity";
        final int REQUEST_OSMAND_API = 1001;
        final int REQUEST_NAVIGATE_GPX_RAW_DATA = 1002;
        final int REQUEST_SHOW_GPX_RAW_DATA = 1003;
        final int REQUEST_NAVIGATE_GPX_URI = 1004;
        final int REQUEST_SHOW_GPX_URI = 1005;
        final int REQUEST_SHOW_GPX_RAW_DATA_AIDL = 1006;
        final int REQUEST_SHOW_GPX_URI_AIDL = 1007;
        final int REQUEST_NAVIGATE_GPX_RAW_DATA_AIDL = 1008;
        final int REQUEST_NAVIGATE_GPX_URI_AIDL = 1009;
        final int REQUEST_GET_GPX_BITMAP_URI_AIDL = 1010;
        final int REQUEST_COPY_FILE = 1011;
        final int REQUEST_IMPORT_FILE = 1012;
        final String AUTHORITY = "net.osmand.osmandapidemo.fileprovider";
        final String GPX_FILE_NAME = "test.gpx";
        final String SQLDB_FILE_NAME = "test.sqlitedb";

        final String MAP_LAYER_ID = "layer_1";

        final String KEY_UPDATES_LISTENER = "subscribe_for_updates";
        final String KEY_OSMAND_INIT_LISTENER = "on_osmand_init";
        final String KEY_GPX_BITMAP_LISTENER = "on_bitmap_created";
        final String KEY_NAV_INFO_LISTENER = "on_nav_info_update";
        final String KEY_NAV_VOICE_INFO_LISTENER = "on_nav_voice_info_update";
        final String KEY_CONTEXT_BTN_LISTENER = "on_ctx_btn_click";
        final String KEY_OSMAND_LOGCAT_LISTENER = "osmand_logcat_listener";

        final String DEMO_INTENT_URI = "osmand_api_demo://main_activity";

        private final String APP_MODE_CAR = "car";
        private final String APP_MODE_PEDESTRIAN = "pedestrian";
        private final String APP_MODE_BICYCLE = "bicycle";
        private final String APP_MODE_BOAT = "boat";
        private final String APP_MODE_AIRCRAFT = "aircraft";
        private final String APP_MODE_BUS = "bus";
        private final String APP_MODE_TRAIN = "train";

        private final String SPEED_CONST_KILOMETERS_PER_HOUR = "KILOMETERS_PER_HOUR";
        private final String SPEED_CONST_MILES_PER_HOUR = "MILES_PER_HOUR";
        private final String SPEED_CONST_METERS_PER_SECOND = "METERS_PER_SECOND";
        private final String SPEED_CONST_MINUTES_PER_MILE = "MINUTES_PER_MILE";
        private final String SPEED_CONST_MINUTES_PER_KILOMETER = "MINUTES_PER_KILOMETER";
        private final String SPEED_CONST_NAUTICALMILES_PER_HOUR = "NAUTICALMILES_PER_HOUR";

        private final String METRIC_CONST_KILOMETERS_AND_METERS = "KILOMETERS_AND_METERS";
        private final String METRIC_CONST_MILES_AND_FEET = "MILES_AND_FEET";
        private final String METRIC_CONST_MILES_AND_METERS = "MILES_AND_METERS";
        private final String METRIC_CONST_MILES_AND_YARDS = "MILES_AND_YARDS";
        private final String METRIC_CONST_NAUTICAL_MILES = "NAUTICAL_MILES";

        private final String OSMAND_SHARED_PREFERENCES_NAME = "osmand-api-demo";
/**
        private val appModesAll = null
        private val appModesNone = emptyList<String>()
        private val appModesPedestrian = listOf(APP_MODE_PEDESTRIAN)
        private val appModesPedestrianBicycle = listOf(APP_MODE_PEDESTRIAN, APP_MODE_BICYCLE)
        private val appModesExceptAirBoatDefault = listOf(APP_MODE_CAR, APP_MODE_BICYCLE, APP_MODE_PEDESTRIAN)
 */


    private TextView textView;
    private final int number=0;
    private CanFrame frame;
    private String buff;

    public static final String EXTRA_INITIAL_TAB = "initial_tab";
    public static final int TAB_NOW_PLAYING = 0;
    public static final int TAB_ENGINE = 1;

    @SuppressLint("StaticFieldLeak")
    public static AlpdroidApplication application;

    public static boolean locationPermissionGranted;

    private final byte[] message="{\"bus\":0,\"id\":05ED,\"data\":[00,00,00,00,00,11,22,33]}".getBytes();

    /**
     * The {@link PagerAdapter} that will provide fragments for each of the
     * sections. We use a {@link FragmentPagerAdapter} derivative, which will keep every loaded
     * fragment in memory. If this becomes too memory intensive, it may be best to switch to a {@link
     * FragmentStatePagerAdapter}.
     */
    private SectionsPagerAdapter mSectionsPagerAdapter;

    /** The {@link ViewPager} that will host the section contents. */
    private ViewPager mViewPager;

    private final ScheduledExecutorService executor =
            Executors.newScheduledThreadPool(1);

    double lastLatitude =0;
    double lastLongitude =0;

    private int delay = 5000;

    public static ILogger logger;

    @SuppressLint("NoLoggedException")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        application = (AlpdroidApplication) getApplication();
        logger = application.getLogger();
        logger.d("%s : AlpDroid Ok, OnCreate MainActivity", TAG);
        super.onCreate(savedInstanceState);


         application.startVehicleServices();

         application.alpdroidData = new VehicleServices();

         application.startListenerService();

        setContentView(R.layout.activity_main);

     //   application.mOsmAndHelper = new OsmAndHelper(this, REQUEST_OSMAND_API, this);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

        // Set up the ViewPager with the sections adapter.
        mViewPager = findViewById(R.id.container);
        mViewPager.setAdapter(mSectionsPagerAdapter);

        TabLayout tabLayout = findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(mViewPager);

        // Initial tab may have been specified in the intent.
        int initialTab = getIntent().getIntExtra(EXTRA_INITIAL_TAB, TAB_ENGINE);
        mViewPager.setCurrentItem(initialTab);



        try {
            if (ContextCompat.checkSelfPermission(getApplicationContext(), android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED )
            {
                ActivityCompat.requestPermissions(this, new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, 101);
                locationPermissionGranted=true;
            }
            locationPermissionGranted=true;
        } catch (Exception e)
        {
            locationPermissionGranted=true;
           // e.printStackTrace();
        }

        //important! set your user agent to prevent getting banned from the osm servers

        //5.6 and newer
        Configuration.getInstance().setUserAgentValue(BuildConfig.APPLICATION_ID);
        logger.d("MainActivity is launching : ", TAG);


    }

    String resultCodeStr(int resultCode) {
        switch (resultCode) {
            case Activity.RESULT_OK : return "OK";
            case Activity.RESULT_CANCELED : return "Canceled";
            case Activity.RESULT_FIRST_USER : return "First user";
            case OsmAndHelper.RESULT_CODE_ERROR_UNKNOWN : return "Unknown error";
            case OsmAndHelper.RESULT_CODE_ERROR_NOT_IMPLEMENTED : return "Feature is not implemented";
            case OsmAndHelper.RESULT_CODE_ERROR_GPX_NOT_FOUND : return "GPX not found";
            case OsmAndHelper.RESULT_CODE_ERROR_INVALID_PROFILE : return "Invalid profile";
            case OsmAndHelper.RESULT_CODE_ERROR_PLUGIN_INACTIVE : return "Plugin inactive";
            case OsmAndHelper.RESULT_CODE_ERROR_EMPTY_SEARCH_QUERY : return "Empty search query";
            case OsmAndHelper.RESULT_CODE_ERROR_SEARCH_LOCATION_UNDEFINED : return "Search location undefined";
            case OsmAndHelper.RESULT_CODE_ERROR_QUICK_ACTION_NOT_FOUND : return "Quick action not found";

        }
        return "" + resultCode;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_OSMAND_API) {
            if (data != null) {
                Bundle extras = data.getExtras();
                if (extras != null && extras.size() > 0) {
               //   if (application.isBound()) application.alpdroidServices.fromOsmData(extras);

                }

            }
            //     if (resultCode == RESULT_OK)
            // si cas intermédiaire
            //          super.onActivityResult(requestCode, resultCode, data);
            //     else
           // super.onActivityResult(requestCode, resultCode, data);
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.settings_item) {
            Intent intent = new Intent(getBaseContext(), SettingsActivity.class);
            startActivityForResult(intent, 1);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    public void onResume() {
        super.onResume();

            logger.d(TAG, "Application Service Bound after Resume, need to rebind");

            application.startVehicleServices();


        Configuration.getInstance().load(this, PreferenceManager.getDefaultSharedPreferences(this));

    }

    @Override
    public void onPause() {
        super.onPause();
        logger.d(TAG, "MainActivity Pause");

            logger.d(TAG, "MainActivity need to be bound after Pause");
            application.startVehicleServices();

        Configuration.getInstance().load(this, PreferenceManager.getDefaultSharedPreferences(this));


    }



    @Override
    public void onDestroy()
    {
        application.close();
        super.onDestroy();
    }

    @Override
    public void osmandMissing() {
        logger.d("OsmAND missiong : ", "Wrong version detected ?");
    }


    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to one of the
     * sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentStatePagerAdapter {

        List<Fragment> fragments =
                ImmutableList.of(new NowPlayingFragment(), new EngineDisplay(), new ConfortDisplay(), new ComputerDisplay(), new MapsDisplay(500));


        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            return fragments.get(position);
        }

        @Override
        public int getCount() {
            return fragments.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return getString(R.string.tab_now_playing);
                case 1:
                    return getString(R.string.engine_display);

                case 2:
                   return getString(R.string.confort_display);

                case 3:
                    return getString(R.string.computer_display);
                case 4 :
                    return getString(R.string.maps_display);
            }
            return null;
        }
    }


}

File: app/src/main/java/com/alpdroid/huGen10/ui/MainActivity.kt
package com.alpdroid.huGen10.ui


import android.Manifest
import android.accounts.Account
import android.accounts.AccountManager
import android.annotation.SuppressLint
import android.app.admin.DevicePolicyManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.Color
import android.media.session.MediaSession
import android.os.Build
import android.os.Bundle
import android.os.StrictMode
import android.os.StrictMode.ThreadPolicy
import android.provider.Settings
import android.util.Log
import android.view.Window
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentPagerAdapter
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.preference.PreferenceManager
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.alpdroid.huGen10.*
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.google.common.collect.ImmutableList
import com.google.firebase.FirebaseApp
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlin.time.ExperimentalTime


@OptIn(ExperimentalTime::class)
class MainActivity : FragmentActivity()  {

    private val TAG = "MainActivity"

    lateinit var application: AlpdroidApplication

    private var sharedPreferences: SharedPreferences? = null

    private lateinit var myRef : DatabaseReference
    private lateinit var database:FirebaseDatabase


    companion object {

        const val EXTRA_INITIAL_TAB = "initial_tab"
        const val TAB_NOW_PLAYING = 0
        const val TAB_ENGINE = 1

        var locationPermissionGranted = false

    }


    private lateinit var mSectionsPagerAdapter: SectionsPagerAdapter

    /** The [ViewPager] that will host the section contents.  */
    private lateinit var mViewPager: ViewPager2


    private val backgroundImagePreferenceChangedReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            updateBackgroundImage()
        }
    }

    private val languagePreferenceChangedReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {

            updateLanguage()

        }
    }


    private val updateArduinoChangedReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {

            Log.d(TAG,"receive intent to update arduino")

        }
    }


    fun getGoogleAccountName(context: Context): String? {
        val accountManager = AccountManager.get(context)
        val googleAccountType = "com.google"

        val accounts = accountManager.getAccountsByType(googleAccountType)
        if (accounts.isNotEmpty()) {
            // Assuming the first Google account found is the one you want
            val googleAccount = accounts[0] as Account
            return googleAccount.name
        }

        return "not found"
    }

     private fun logAppInstallEvent(uniqueId: String) {


         val devicePolicyManager =
             this@MainActivity.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager

         val androidId = Settings.Secure.getString(
             this.contentResolver,
            Settings.Secure.ANDROID_ID
         )


         val hardwareInfo = "Model: ${Build.MODEL}, Product: ${Build.PRODUCT}, Hardware: ${Build.HARDWARE}"

         val versionCode=BuildConfig.VERSION_CODE

         myRef = database.getReference("MyAlpDroid Installation - $versionCode")

         myRef.child("UniqueID-$androidId").child("Version").setValue(BuildConfig.VERSION_NAME)
         myRef.child("UniqueID-$androidId").child("hardwareKey").setValue(hardwareInfo)
         myRef.child("UniqueID-$androidId").child("dateformat").setValue(uniqueId)
         myRef.child("UniqueID-$androidId").child("accountName").setValue(getGoogleAccountName(application))

         myRef.child("UniqueID-$androidId").child("num_install").get().addOnSuccessListener {
             if (it.value!=null)
                myRef.child("UniqueID-$androidId").child("num_install").setValue(it.value.toString().toInt()+1)
             else
                 myRef.child("UniqueID-$androidId").child("num_install").setValue(1.toString())
         }.addOnFailureListener{
             Log.d(TAG, "Failure data Firebase")
         }



     }

    fun getCurrentDateFormatted(): String {
        val currentTimeMillis = System.currentTimeMillis()
        val dateFormat = SimpleDateFormat("ddMMyy", Locale.getDefault())
        val date = Date(currentTimeMillis)
        return dateFormat.format(date)
    }

    @SuppressLint("NoLoggedException")
    override fun onCreate(savedInstanceState: Bundle?) {


        super.onCreate(savedInstanceState)


        application=getApplication() as AlpdroidApplication

        AlpdroidApplication.setContext(application)

        FirebaseApp.initializeApp(this)

        database = FirebaseDatabase.getInstance()

        logAppInstallEvent(getCurrentDateFormatted())

        AlpdroidApplication.controller = MediaSession(application, TAG).controller

        application.startVehicleServices()
        application.alpdroidData = VehicleServices()
        application.startListenerService()

        val oldPolicy: ThreadPolicy
        oldPolicy = StrictMode.getThreadPolicy()
        StrictMode.allowThreadDiskReads()
        try {
            // Do reads here
           setContentView(R.layout.activity_main)
        } finally {
            StrictMode.setThreadPolicy(oldPolicy)
        }


        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        mSectionsPagerAdapter = SectionsPagerAdapter(this)

        // Set up the ViewPager2 with the sections adapter.
        mViewPager = findViewById(R.id.container)
        mViewPager.adapter=mSectionsPagerAdapter

        val tabLayout = findViewById<TabLayout>(R.id.tabs)
        TabLayoutMediator(tabLayout, mViewPager) { tab, position ->
            tab.text = mSectionsPagerAdapter.getPageTitle(position+1)
        }.attach()

        // Initial tab may have been specified in the intent.
        val initialTab = intent.getIntExtra(EXTRA_INITIAL_TAB, TAB_ENGINE)
        mViewPager.setCurrentItem(initialTab)

        try {
            if (ContextCompat.checkSelfPermission(
                    applicationContext,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                    ),
                    101
                )
                locationPermissionGranted = true
            }
            locationPermissionGranted = true
        } catch (e: Exception) {
            locationPermissionGranted = true
            // e.printStackTrace();
        }

        updateBackgroundImage()



        StrictMode.allowThreadDiskReads()
        try {
              sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        } finally {
            StrictMode.setThreadPolicy(oldPolicy)
        }

        if (sharedPreferences!!.getBoolean(getString(R.string.isdownload_UNO),false))
        {
            Toast.makeText(this, "Arduino code updated",5.toInt()).show()
            sharedPreferences!!.edit().putBoolean(getString(R.string.isdownload_UNO),false).apply()
        }

        if (application.alpdroidServices.isArduinoOpen())
        {
            Toast.makeText(this, "Arduino is Open",20.toInt()).show()
        }
        else
            Toast.makeText(this, "Arduino is not Up",20.toInt()).show()

    }

    fun setLocale(languageCode: String) {

        val config = resources.configuration

        val locale = Locale(languageCode)
        Locale.setDefault(locale)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1)
            config.setLocale(locale)
        else
            config.locale = locale

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            createConfigurationContext(config)
        resources.updateConfiguration(config, resources.displayMetrics)
    }

    private fun updateLanguage()
    {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)

        val languetoggle = sharedPreferences.getBoolean("Langue", false)

        if (languetoggle)
            setLocale("en")
        else
            setLocale("fr")

        val refresh = Intent(this, MainActivity::class.java)
        startActivity(refresh)

    }
    private fun updateBackgroundImage() {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)

        var backgroundImage = "background"
        var taskbarColor = "#000000"

        val string = sharedPreferences.getString("Choix", "background,#031627")
        val targetChar = ','
        val targetIndex = string?.indexOf(targetChar)!!+1

        if (targetIndex>0)
     {
             backgroundImage = string.substring(0, targetIndex - 1)
             taskbarColor = (string.substring(targetIndex))
        }


        val backgroundImageResource = resources.getIdentifier(backgroundImage, "drawable", packageName)

        mViewPager.setBackgroundResource(backgroundImageResource)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window: Window = window
            window.statusBarColor = Color.parseColor(taskbarColor)
        }
    }


    public override fun onRestart() {

   //     MLog.d(TAG, "Activity onRestart : Begin")
        application=getApplication() as AlpdroidApplication

        AlpdroidApplication.setContext(application)

        application.startVehicleServices()
        application.alpdroidData = VehicleServices()
        application.startListenerService()


        super.onRestart()

        //MLog.d(TAG, "Activity onRestart : End")

    }


    public override fun onResume() {
        val oldPolicy: ThreadPolicy
      //  MLog.d(TAG, "Activity onResume : Begin")
        super.onResume()


        oldPolicy = StrictMode.getThreadPolicy()
        StrictMode.allowThreadDiskReads()
        StrictMode.allowThreadDiskWrites()
        try {
            // Do reads here
            //Configuration.getInstance().load(this, PreferenceManager.getDefaultSharedPreferences(this));
        } finally {
            StrictMode.setThreadPolicy(oldPolicy)
        }
        //MLog.d(TAG, "Activity onResume : End")
        LocalBroadcastManager.getInstance(this).registerReceiver(backgroundImagePreferenceChangedReceiver, IntentFilter("change_background"))
        LocalBroadcastManager.getInstance(this).registerReceiver(languagePreferenceChangedReceiver, IntentFilter("change_language"))

    }

    public override fun onPause() {
        val oldPolicy: ThreadPolicy
     //   MLog.d(TAG, "Activity onPause : Begin")
        super.onPause()

        LocalBroadcastManager.getInstance(this).unregisterReceiver(backgroundImagePreferenceChangedReceiver)
        LocalBroadcastManager.getInstance(this).unregisterReceiver(languagePreferenceChangedReceiver)

        //  application.startVehicleServices();
        //  application.alpdroidData = new VehicleServices();
        oldPolicy = StrictMode.allowThreadDiskReads()
        try {
            // Do reads here
            //Configuration.getInstance().load(this, PreferenceManager.getDefaultSharedPreferences(this));
        } finally {
            StrictMode.setThreadPolicy(oldPolicy)
        }
     //   MLog.d(TAG, "Activity onPause : End")
    }

    public override fun onDestroy() {
        //MLog.d(TAG, "Activity onDestroy : Begin")
        super.onDestroy()

    }


    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to one of the
     * sections/tabs/pages.
     */
    inner class SectionsPagerAdapter(fm: FragmentActivity?) :
        FragmentStateAdapter(fm!!) {
        @OptIn(ExperimentalStdlibApi::class, ExperimentalUnsignedTypes::class)
        var fragments: List<Fragment> = ImmutableList.of(
            NowPlayingFragment(),
            EngineDisplay(),
            ConfortDisplay(),
            ComputerDisplay(),
           BargrapheFragment(),
            RoadBook(),
            SettingsDisplay(),
            SimpleCounterDisplay(),
        )

        override fun createFragment(position: Int): Fragment {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).

            return fragments[position]
        }

        override fun getItemCount(): Int {
            return fragments.size
        }

        fun getPageTitle(position: Int): CharSequence? {
            when (position) {
                0 -> return getString(R.string.tab_now_playing)
                1 -> return getString(R.string.engine_display)
                2 -> return getString(R.string.confort_display)
                3 -> return getString(R.string.computer_display)
                4 -> return "Télémétrie"
                4 -> return "RoadBook"
                5 -> return getString(R.string.settings)
                6 -> return "Simple"
            }
            return null
        }
    }




}